From c5f2977c8bcc0a3c4c466af1984a23de35effe09 Mon Sep 17 00:00:00 2001
From: Zubair Lutfullah <zubair.lutfullah@gmail.com>
Date: Thu, 11 Jul 2013 19:36:34 +0100
Subject: [PATCH 05/19] input: ti_tsc: Enable shared IRQ for TSC

Touchscreen and ADC share the same IRQ line from parent

MFD core.

Previously only Touchscreen was interrupt based.

With continuous mode support added in ADC driver, now driver requires

interrupt to process the ADC samples, so enable shared IRQ flag bit for

touchscreen.

Forward ported from Arago tree
---
 drivers/input/touchscreen/ti_am335x_tsc.c |   16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/input/touchscreen/ti_am335x_tsc.c b/drivers/input/touchscreen/ti_am335x_tsc.c
index 0e9f02a..cf56cae 100644
--- a/drivers/input/touchscreen/ti_am335x_tsc.c
+++ b/drivers/input/touchscreen/ti_am335x_tsc.c
@@ -260,8 +260,16 @@ static irqreturn_t titsc_irq(int irq, void *dev)
 	unsigned int fsm;
 
 	status = titsc_readl(ts_dev, REG_IRQSTATUS);
-	if (status & IRQENB_FIFO0THRES) {
-
+ 
+        /*
+         * ADC and touchscreen share the IRQ line.
+         * FIFO1 threshold interrupt is used by ADC,
+         * hence return from touchscreen IRQ handler if FIFO1
+         * threshold interrupt occurred.
+         */
+        if (status & IRQENB_FIFO1THRES)
+                return IRQ_NONE;
+        else if (status & IRQENB_FIFO0THRES) {
 		titsc_read_coordinates(ts_dev, &x, &y, &z1, &z2);
 
 		if (ts_dev->pen_down && z1 != 0 && z2 != 0) {
@@ -315,7 +323,7 @@ static irqreturn_t titsc_irq(int irq, void *dev)
 	}
 
 	if (irqclr) {
-		titsc_writel(ts_dev, REG_IRQSTATUS, irqclr);
+		titsc_writel(ts_dev, REG_IRQSTATUS, (status | irqclr) );
 		am335x_tsc_se_update(ts_dev->mfd_tscadc);
 		return IRQ_HANDLED;
 	}
@@ -389,7 +397,7 @@ static int titsc_probe(struct platform_device *pdev)
 	}
 
 	err = request_irq(ts_dev->irq, titsc_irq,
-			  0, pdev->dev.driver->name, ts_dev);
+			  IRQF_SHARED, pdev->dev.driver->name, ts_dev);
 	if (err) {
 		dev_err(&pdev->dev, "failed to allocate irq.\n");
 		goto err_free_mem;
-- 
1.7.9.5

